### PR TITLE
Pallet grouping macros

### LINKED ISSUE #3: System grouping
includes the System pallet abstraction group

### LINKED ISSUE #5: Finalize the API design and showcase it on readme
Readme should clearly convey:
- the problem we are solving
- how are we solving it
    - the API we are proposing

### LINKED ISSUE #10: Consensus grouping
- pallet_authorship
- pallet_collator_selection
- pallet_session
- pallet_aura
- cumulus_pallet_aura_ext

### LINKED ISSUE #11: Asset grouping
- pallet_transaction_payment
- pallet_assets
- pallet_asset_manager

### LINKED ISSUE #12: EVM grouping
- pallet_ethereum
- pallet_evm
- pallet_base_fee
- pallet_evm_chain_id

### LINKED ISSUE #13: Governance grouping
- pallet_sudo
- pallet_conviction_voting
- pallet_referenda
- pallet_custom_origins
- pallet_whitelist
- pallet_treasury

### LINKED ISSUE #14: Xcm grouping
- cumulus_pallet_xcmp_queue
- pallet_xcm
- cumulus_pallet_xcm
- pallet_message_queue

### PR DESCRIPTION
**Pull request by peektism**
Implements wrapper macros for all pallet groupings

**[PR using macros in both runtime templates](https://github.com/OpenZeppelin/polkadot-runtime-templates/pull/300)**

- [x] closes #3 
- [x] closes #10 
- [x] closes #13 
- [x] closes #14 
- [x] closes #11 
- [x] closes #12 

Final TODOs:
- [x] in-line docs per grouping (thanks @ggonzalez94 )
- [x] closes #5 (done in https://github.com/OpenZeppelin/polkadot-runtime-wrappers/pull/1/commits/640268ec44273fce4fc2b9549455b85caa716494 https://github.com/OpenZeppelin/polkadot-runtime-wrappers/pull/1/commits/07ce4fb52694e21696f7f349cdcb65f347f93d22)

Tanssi fixes:
- [x] move pallet-sessions to consensus grouping from system grouping
- [x] expose ProxyType in the system grouping

Patches merged into macros that should be reviewed closely:
- [x] https://github.com/OpenZeppelin/polkadot-runtime-templates/pull/331
- [x] https://github.com/OpenZeppelin/polkadot-runtime-templates/pull/343
- [x] https://github.com/OpenZeppelin/polkadot-runtime-templates/pull/345 
- [x] https://github.com/OpenZeppelin/polkadot-runtime-templates/pull/346 
- [x] https://github.com/OpenZeppelin/polkadot-runtime-templates/pull/322
- [x] https://github.com/OpenZeppelin/polkadot-runtime-templates/pull/327
- [x] https://github.com/OpenZeppelin/polkadot-runtime-templates/pull/333
- [x] https://github.com/OpenZeppelin/polkadot-runtime-templates/pull/341

## Follow ups
- #8
- #24
- #25
- #26
- #27
- #28
- #29

### CODE REVIEW

**Comment by KitHat** [comment]
@4meta5 
I have looked through the groupings and haven't seen `pallet_asset_manager` in neither of abstractions.
* We need it in both of our templates (Generic will need it with introduction of Snowbridge support)
* It will affect the XCM benchmarking -- we need to register assets in AssetManager, and this benchmark should be generated by our macro.
Could you please add it?

**Comment by ozgunozerk** [comment]
> yes because it has not been configured in the generic template yet, will open an issue for this but we should configure them in both templates before adding it to the macros

I configured it due to it was necessary also for snowbridge integration. If it is not there, it is probably on one of my branches, will push soon

**Comment by ozgunozerk** [comment]
Ok I found it, however it is intertwined with `orml_xtokens` config. It's late for me, so I can tackle that tomorrow. 

If you do not want to wait, the branch is `ozgun-xtokens-on-generic`.

Let me know if you will extract it yourself or you are ok with waiting for tomorrow @4meta5

**Review by ggonzalez94 — COMMENTED** [review]
Looking good! Just did a first pass and will continue tomorrow

File: src/xcm.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+        impl cumulus_pallet_xcmp_queue::Config for Runtime {
+            type ChannelInfo = ParachainSystem;
+            type ControllerOrigin = <$t as XcmConfig>::XcmpQueueControllerOrigin;
+            type ControllerOriginConverter = XcmOriginToTransactDispatchOrigin;
+            type MaxActiveOutboundChannels = <$t as XcmConfig>::MaxActiveOutboundChannels;
+            type MaxInboundSuspended = <$t as XcmConfig>::XcmpQueueMaxInboundSuspended;
+            type MaxPageSize = <$t as XcmConfig>::MaxPageSize;
+            /// Ensure that this value is not set to null (or NoPriceForMessageDelivery) to prevent spamming
+            type PriceForSiblingDelivery = PriceForSiblingParachainDelivery;
```
Comment (KitHat) [inline src/xcm.rs:45]: Shouldn't this type be passed from somewhere or defined? I can't see it defined as a field of macro or by macro.

Comment (KitHat) [reply src/xcm.rs:45]: Moreover, when people will try to use this macro, they will receive an error about a missing import and won't be able to find a documentation about it.

File: src/xcm.rs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            type RuntimeEvent = RuntimeEvent;
+            type VersionWrapper = ();
+            /// Rerun benchmarks if you are making changes to runtime configuration.
+            type WeightInfo = weights::cumulus_pallet_xcmp_queue::WeightInfo<Runtime>;
```
Comment (KitHat) [inline src/xcm.rs:49]: This is a reference to the module, that is not generated by macro and dependency will fall out of nowhere on a macro user. This type should be provided from macro invocation.

File: src/xcm.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            // Enqueue XCMP messages from siblings for later processing.
+            type XcmpQueue =
+                TransformOrigin<MessageQueue, AggregateMessageOrigin, ParaId, ParaIdToSibling>;
```
Comment (KitHat) [inline src/xcm.rs:52]: Also, if a macro uses some external type, it should import it as well.
Let's even think about re-exporting the crates like `frame_support` and others that we actively use..

**Comment by KitHat** [comment]
I have added a couple of comments that are quite crucial for DX in my opinion. We should discuss them together and maybe resolve in the next release, but leaving some types and modules undefined yet used and imports unresolved is a bad idea.

File: Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "polkadot-runtime-wrappers"
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply Cargo.toml:2]: Maybe we should have... but this one will be imported by our templates and other templates, so I think here it is more important. I'm happy to go with `openzeppelin-polkadot-wrappers` if you like that one better.

File: src/system.rs [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+                        c,
+                        RuntimeCall::Proxy(pallet_proxy::Call::reject_announcement { .. })
+                            | RuntimeCall::Multisig { .. }
+                    ),
+                    ProxyType::Collator => {
+                        matches!(c, RuntimeCall::CollatorSelection { .. } | RuntimeCall::Multisig { .. })
+                    }
+                }
+            }
+        }
```
Comment (KitHat) [inline src/system.rs:201]: There is a problem that introduces coupling between system and consensus macros.
`ProxyType::Collator` assumes that there is a `CollatorSelection`-named pallet implemented on the chain. It is not guaranteed. I understand that `Balances` and `Multisig` are also not guaranteed but IMO `CollatorSelection` is too much of an assumption. 

However, I do not know a better solution.

**Comment by ozgunozerk** [comment]
> There are many cases (100+) in addition to the comments made where user errors significantly improve if we include imports locally.

Just want to state this fact in case it might be helpful for this convo:

Generally, in macros, it is bad practice to have `import` statements due to macro hygiene. 
The reason is: the user may have the same import statement in his code that is outside of the macro's scope. Then, they will have conflicting import statement. 

To prevent that, it is usually preferred to have the items with full path in your macros. 

For example, instead of:
```rust
import sp_core::H256;

pub type Hash = H256;
```

the macro should have:
```rust
pub type Hash = sp_core::H256;
```

File: Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+version = "0.0.1"
+
+[dependencies]
+frame-support = { git = "https://github.com/paritytech/polkadot-sdk", default-features = false, tag = "polkadot-stable2407-1" }
+scale-info = { version = "2.11.1", default-features = false }
+sp-std = { git = "https://github.com/paritytech/polkadot-sdk", default-features = false, tag = "polkadot-stable2407-1" }
+sp-version = { git = "https://github.com/paritytech/polkadot-sdk", default-features = false, tag = "polkadot-stable2407-1" }
+
+[features]
+default = [ "std" ]
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply Cargo.toml:21]: yes, I understand all of this - I just thought that if we expect most of the usage to come from `no_std` parachains then we shouldn't make `std` the default, instead it should be opt in. But not a major thing at this point, I can create an issue to think about this in the future if you agree there's value

File: Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "openzeppelin-polkadot-wrappers"
+authors = [ "OpenZeppelin" ]
+description = "Wrappers with Sensible Defaults for constructing Polkadot Parachain Runtimes"
```
Comment (ggonzalez94) [inline Cargo.toml:4]: Nit:
```suggestion
description = "Wrappers with sensible defaults for constructing Polkadot Parachain Runtimes"
```

File: Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+edition = "2021"
+license = "GPL-3.0-only"
+repository = "https://github.com/OpenZeppelin/polkadot-runtime-wrappers"
+readme = "README.md"
+version = "0.0.1"
```
Comment (ggonzalez94) [inline Cargo.toml:9]: Don't you think 0.1.0 is more appropiate? Still signals that there will be breaking changes, but gives is the opportunity to issue patch releases(e.g. 0.1.1, 0.1.2, etc.)
```suggestion
version = "0.1.0"
```

File: src/assets.rs [hunk 9]
```diff
@@ -0,0 +1,68 @@
+//! Assets pallet groupings wrapper
```
Comment (ggonzalez94) [inline src/assets.rs:1]: I think we should be doing a better job at code documentation(feel free to incorporate these suggestions in a separate PR if you think that's easier). I'm putting an example for assets, but we should do it for all the others as well. Also adding an example would be great, if possible

```suggestion
//// Implements the OpenZeppelin assets configuration for a Runtime.
///
/// This macro sets up the necessary configurations for the following pallets:
/// - `pallet_assets`
/// - `pallet_transaction_payment`
/// - `pallet_asset_manager`
///
/// # Parameters
/// - `$t`: A type that implements the `AssetsConfig` trait, providing the necessary associated types
///   and configurations.
///
/// # Note
/// Ensure your type implements the `AssetsConfig` trait with all required associated types:
/// - `ApprovalDeposit`
/// - `AssetDeposit`
/// - `AssetId`
/// - `CreateOrigin`
/// - `ForceOrigin`
/// - `WeightToFee`
/// - `AssetRegistrar`
/// - `AssetRegistrarMetadata`
/// - `AssetType`
/// - `ForeignAssetModifierOrigin`
///
/// # Important
/// Rerun benchmarks if making changes to runtime configuration, as weight calculations
/// may need to be updated.
```

File: src/assets.rs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            /// Relay Chain `TransactionByteFee` / 10
+            pub const TransactionByteFee: Balance = 10 * MICROCENTS;
+            pub const OperationalFeeMultiplier: u8 = 5;
+        }
+
+        impl pallet_transaction_payment::Config for Runtime {
+            /// There are two possible mechanisms available: slow and fast adjusting.
+            /// With slow adjusting fees stay almost constant in short periods of time, changing only in long term.
+            /// It may lead to long inclusion times during spikes, therefore tipping is enabled.
+            /// With fast adjusting fees change rapidly, but fixed for all users at each block (no tipping)
```
Comment (ggonzalez94) [inline src/assets.rs:47]: why are all of these /// comments? I don't think these are intended to be rendered in docs.rs or similar right?

**Review by ggonzalez94 — COMMENTED** [review]
Left a few nits and suggestions, especially around comments and how we document the macro. I think:
- We should step up our game in terms of in-line docs, each grouping should have an explanation of what it does and what pallets it uses(I tried to provide an example for assets)
- I saw a lot of inconsistencies in comments(/// and //), these were probably porter over from the templates repo, but I think it is a good opportunity to fix them.

**Review by KitHat — APPROVED** [review]

File: src/consensus.rs [hunk 11]
```diff
@@ -0,0 +1,80 @@
+//! Consensus pallet grouping wrapper
```
Comment (ggonzalez94) [inline src/consensus.rs:1]:
```suggestion
//! This macro sets up the necessary configurations for the following pallets:
//! - `pallet_authorship`
//! - `pallet_aura`
//! - `cumulus_pallet_aura_ext`
//! - `pallet_collator_selection`
//! - `pallet_session`
//!
//! # Parameters
//! - `$t`: A type that implements the `ConsensusConfig` trait, providing the necessary associated types
//!   and configurations.
//!
//! # Important
//! Rerun benchmarks if making changes to runtime configuration, as weight calculations
//! may need to be updated.
```

File: src/evm.rs [hunk 12]
```diff
@@ -0,0 +1,80 @@
+#[macro_export]
```
Comment (ggonzalez94) [inline src/evm.rs:1]:
```suggestion
//! Implements the OpenZeppelin EVM configuration for a Runtime.
//!
//! This macro sets up the necessary configurations for the following pallets:
//! - `pallet_ethereum`
//! - `pallet_evm`
//! - `pallet_evm_chain_id`
//! - `pallet_base_fee`
//!
//! # Parameters
//! - `$t`: A type that implements the `EvmConfig` trait, providing the necessary associated types
//!   and configurations.
//!
//! # Important
//! Rerun benchmarks if making changes to runtime configuration, as weight calculations
//! may need to be updated.

#[macro_export]
```

File: src/governance.rs [hunk 13]
```diff
@@ -0,0 +1,100 @@
+//! Governance pallet groupings wrapper
```
Comment (ggonzalez94) [inline src/governance.rs:1]:
```suggestion
//! Implements the OpenZeppelin governance configuration for a Runtime.
//!
//! This macro sets up the necessary configurations for the following pallets:
//! - `pallet_sudo`
//! - `pallet_treasury`
//! - `pallet_conviction_voting`
//! - `pallet_whitelist`
//! - `pallet_custom_origins`
//! - `pallet_referenda`
//!
//! # Parameters
//! - `$t`: A type that implements the `GovernanceConfig` trait, providing the necessary associated types
//!   and configurations.
//!
//! # Important
//! Rerun benchmarks if making changes to runtime configuration, as weight calculations
//! may need to be updated.
```

File: src/system.rs [hunk 14]
```diff
@@ -0,0 +1,243 @@
+//! System pallet groupings wrapper
```
Comment (ggonzalez94) [inline src/system.rs:1]:
```suggestion
//! Implements the OpenZeppelin system configuration for a Runtime.
//!
//! This macro sets up the necessary configurations for the following pallets:
//! - `frame_system`
//! - `pallet_timestamp`
//! - `parachain_info`
//! - `pallet_scheduler`
//! - `pallet_preimage`
//! - `pallet_proxy`
//! - `pallet_balances`
//! - `pallet_utility`
//! - `cumulus_pallet_parachain_system`
//! - `pallet_multisig`
//!
//! # Parameters
//! - `$t`: A type that implements the `SystemConfig` trait, providing the necessary associated types
//!   and configurations for core system functionality.
//!
//! # Important
//! Rerun benchmarks if making changes to runtime configuration, as weight calculations
//! may need to be updated.
```

File: src/xcm.rs [hunk 15]
```diff
@@ -0,0 +1,242 @@
+//! XCM pallet groupings wrapper
```
Comment (ggonzalez94) [inline src/xcm.rs:1]:
```suggestion
//! Implements the OpenZeppelin XCM configuration for a Runtime.
//!
//! This macro sets up the necessary configurations for the following pallets:
//! - `pallet_message_queue`
//! - `cumulus_pallet_xcmp_queue`
//! - `pallet_xcm`
//! - `cumulus_pallet_xcm`
//! - `pallet_xcm_weight_trader`
//! - `orml_xtokens`
//! - `pallet_xcm_transactor`
//!
//! # Parameters
//! - `$t`: A type that implements the `XcmConfig` trait, providing the necessary associated types
//!   and configurations for cross-chain messaging functionality.
//!
//! # Important
//! Rerun benchmarks if making changes to runtime configuration, as weight calculations
//! may need to be updated.
```

**Review by ggonzalez94 — APPROVED** [review]
Looks good! Left some suggestions for the module comments, feel free to take them if you didn't work on them already

**Pull request merged**